scheduler/reconciler/reconcile_node.go

```diff
@@ -68,21 +75,54 @@ func (nr *NodeReconciler) Compute(
 	return result
 }
 
-// diffSystemAllocsForNode is used to do a set difference between the target allocations
-// and the existing allocations for a particular node. This returns 8 sets of results,
-// the list of named task groups that need to be placed (no existing allocation), the
-// allocations that need to be updated (job definition is newer), allocs that
-// need to be migrated (node is draining), the allocs that need to be evicted
-// (no longer required), those that should be ignored, those that are lost
-// that need to be replaced (running on a lost node), those that are running on
-// a disconnected node but may resume, and those that may still be running on
-// a node that has resumed reconnected.
+// computeCanaryNodes is a helper function that, given required task groups and
+// eligible nodes, outputs a map of arrays of nodes, indexed by task group name,
+// on which canaries should be placed.
+func computeCanaryNodes(required map[string]*structs.TaskGroup,
+	eligibleNodes map[string]*structs.Node) map[string][]*structs.Node {
+
+	canaryNodes := map[string][]*structs.Node{}
+	eligibleNodesList := slices.Collect(maps.Values(eligibleNodes))
+	for _, tg := range required {
+		if tg.Update.IsEmpty() || tg.Update.Canary == 0 {
+			continue
+		}
+
+		for i, n := range eligibleNodesList {
+			if i > (tg.Update.Canary*len(eligibleNodes))/100 {
+				break
+			}
+
+			if canaryNodes[tg.Name] == nil {
+				canaryNodes[tg.Name] = []*structs.Node{}
+			}
```

Comment:
Nitpick: you can probably make this unconditional before you start the loop over `eligibleNodes` list, because we'll never have 0% at that point and always have at least one node in the list.

Reply:
[1508d17](https://github.com/hashicorp/nomad/pull/26499/commits/1508d17efe2dcb48047bbc2ffefa13db9f22ad05)